plugins/org.eclipse.cbi.p2repo.sbom/src/org/eclipse/cbi/p2repo/sbom/SBOMApplication.java

```diff
@@ -704,6 +711,38 @@ public IStatus run(IProgressMonitor monitor) throws ProvisionException {
 			return Status.OK_STATUS;
 		}
 
+		private void loadArtifactSource(URI location, URI referenced, Set<URI> loaded) {
+			if (loaded.add(location)) {
+				try {
+					var artifactManager = getArtifactRepositoryManager();
+					var contains = artifactManager.contains(location);
+					var repository = artifactManager.loadRepository(location, new NullProgressMonitor());
+					artifactSourceRepositories
+							.add(new ArtifactSourceRepository(referenced == null ? location : referenced, repository));
+					if (!contains) {
+						artifactManager.removeRepository(location);
+					}
+					var metadataManager = getMetadataRepositoryManager();
+					var references = metadataManager.loadRepository(location, new NullProgressMonitor())
+							.getReferences();
+					contains = metadataManager.contains(location);
```

Comment:
These two are in the wrong order.   I'm also kind of concerned that if there are composites involved that do atomic loading those children might be added already too.